File: Sorting-Algorithms/6-Heap Sort/Heap Sort.py
#-----------------------------------------------------------------
#                   HEAP SORT
#-----------------------------------------------------------------

def heapify(arr, n, i):
    """
        heapify function ==> preserves the maximum heap property in the root subtree i.
        Assumes that the left and right subtrees are already heaps.
        arr: list of numbers
        n: heap size (number of valid elements in the array)
        i: index of the root of the subtree
    """
    largest = i          # Meaning ==> Root is Biggest
    left = 2 * i + 1     # Left child index
    right = 2 * i + 2    # Right child index

    # If Left child Bigger than Right child
    if left < n and arr[left] > arr[largest]:
        largest = left

    # If Right child Bigger than Left child
    if right < n and arr[right] > arr[largest]:
        largest = right

    # If the right child is greater than the largest current
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]   # Swap
        heapify(arr, n, largest)   #  Heapify on Subtree is changed

def build_heap(arr):
    
    n = len(arr)
    
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

def heap_sort(arr):
    
    n = len(arr)

    # Step 1 ==> Build the maximum heap
    build_heap(arr)

    # Step 2 ==> Extract elements from the heap one by one
    for i in range(n - 1, 0, -1):
        # Swap the root (largest element) with the last element
        arr[0], arr[i] = arr[i], arr[0]
        # heapify on the root by reducing the heap size (i)
        heapify(arr, i, 0)


def print_array(arr):
    print(" ".join(map(str, arr)))

# Main
if __name__ == "__main__":
    arr = [64, 45, 48, 87, 878, 78, 74, 13, 55, 44, 34, 25, 12, 22, 100, 210, 150, 124, 11, 90]
    print(" BEFORE SORTING : ")
    print_array(arr)

    heap_sort(arr)

    print(" AFTER SORTING : ")
    print_array(arr)

File: Sorting-Algorithms/6-Heap Sort/Heap Sort.cpp
//=================================================
//              HEAP SORT
//=================================================

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// n = Heap Length
// heapify function: preserve the maximum heap property in the root subtree i
void heapify(vector<int>& arr, int n, int i)
{
    int largest = i;    // Meaning ==> Root is Biggest
    int left = 2 * i + 1;   // Left child index
    int right = 2 * i + 2;  // Right child index


    // If Left child Bigger than Right child
    if(left < n && arr[left] > arr[largest])
        largest = left;

    // If Right child Bigger than Left child

    if (right < n && arr[right] > arr[largest])
        largest = right;

    // If the right child is greater than the largest current
    if(largest != i)
    {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);       // Heapify on Subtree is changed
    }
}


// Create a max heap from the input array
void buildHeap(vector<int>& arr)
{
    int n = arr.size();
    // Find the last non-leaf node: (n/2 - 1)
    // and heapify from there
    for(int i = n / 2 - 1; i >= 0; i++)
    {
        heapify(arr, n, i);
    }
}


// Heap Sorting
void heapSort(vector<int>& arr)
{
    int n = arr.size();

    // Step 1 ==> Build the maximum heap
    buildHeap(arr);

    // Step 2 ==> Extract elements from the heap one by one
    for(int i = n - 1; i > 0; i--)
    {
        // Swap the root (largest element) with the last element
        swap(arr[0], arr[i]);

        // heapify on the root by reducing the heap size (i)
        heapify(arr, i, 0);
    }
}


void printArray(const vector<int>& arr)
{
    for(int val : arr)
        cout << val << " ";
    cout << endl;
}


// Main Func
int main()
{
    vector<int> arr = {64, 45, 48, 87, 878, 78, 74, 13, 55, 44, 34, 25, 12, 22, 100, 210, 150, 124, 11, 90};

    cout << " BEFORE SORTING : ";
    printArray(arr);

    heapSort(arr);

    cout << " AFTER SORTING : ";
    printArray(arr);

    return 0;
}
